src/Weather/Skysight/CDFDecoder.cpp

```diff
@@ -0,0 +1,334 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2016 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "CDFDecoder.hpp"
+
+#ifdef ANDROID
+#include <netcdfcpp.h>
+#include <geotiffio.h>
+#include <xtiffio.h>
+#else
+#include <netcdf>
+#include <geotiff/geotiffio.h>
+#include <geotiff/xtiffio.h>
+#endif
+
+#include "SkysightAPI.hpp"
+#include "OS/FileUtil.hpp"
+#include "LogFile.hpp"
+
+void CDFDecoder::DecodeAsync()
+{
+  std::lock_guard<Mutex> lock(mutex);
+
+  Trigger();
+}
+
+void CDFDecoder::DecodeSync()
+{
+  Decode();
+}
+
+void CDFDecoder::Done()
+{
+  StandbyThread::LockStop();
+}
+
+CDFDecoder::Status CDFDecoder::GetStatus()
+{
+  std::lock_guard<Mutex> lock(mutex);
+  CDFDecoder::Status s = status;
+  return s;
+}
+
+void CDFDecoder::Tick() noexcept
+{
+  status = Status::Busy;
+  mutex.unlock();
+  Decode();
+}
+
+bool CDFDecoder::DecodeError() {
+  LogFormat("Decoder Error");
+  MakeCallback(false);
+  mutex.lock();
+  status = Status::Error;
+  return true;
+}
+
+bool CDFDecoder::DecodeSuccess()
+{
+  MakeCallback(true);
+  mutex.lock();
```

Comment:
Never use unprotected mutex lock/unlock calls!

Reply:
mutex which guards status in line 83 is unlocked on line 84 in the next patch.